### ISSUE TITLE
[ORM] Inconsitent implementations of SS_Limitable

### ISSUE THREAD

**Issue by Zauberfisch** [silverstripe/silverstripe-framework#2949]
`DataList` and `ArrayList` implement the `SS_Limitable` interface. they however have inconsistent behavior in `$list->limit()` when passing anything other than positive integer values.

```
$arrayList->limit(-1); // returns all records except the last one, so its $l->limit($l->count()-1)
$dataList->limit(-1); // throws InvalidArgumentException

$arrayList->limit("some string"); // returns 0 records
$dataList->limit("some string"); // raises a '[Notice] Undefined offset: 1' but still returns all records
```

and then there is the fact that `->limit(0)` returns all items, see #2950 for the full issue.

----

## To do

 * [x] Deprecate in SS4 (#8439)
 * [ ] Change ArrayList and DataList behaviour in SS5

**Comment by kinglozzer** [comment]
My vote would be to throw `InvalidArgumentExceptions` for both negative and string values in both `ArrayList` and `DataList`

**Comment by dhensby** [comment]
> My vote would be to throw InvalidArgumentExceptions for both negative and string values in both ArrayList and DataList

only if they fail an `is_numeric` check, though...

**Comment by robbieaverill** [comment]
@sminnee I've merged #8439, do we need to merge it up and have a follow up PR for 5.x? I guess we can use scalar type hints instead of the first couple of if statements in that that PR too

**Comment by sminnee** [comment]
Yeah it's probably a good idea to leave this ticket open until then, yeah.
